contracts/simple_dex/lib.rs

```diff
@@ -0,0 +1,483 @@
+                return Err(DexError::UnsupportedToken(token_out));
+            }
+
+            // check allowance
+            if self.allowance(token_in, caller, this)? < amount_token_in {
+                return Err(DexError::InsufficientAllowanceOf(token_in));
+            }
+
+            let balance_token_in = self.balance_of(token_in, this)?;
+            let balance_token_out = self.balance_of(token_out, this)?;
+
+            if balance_token_out < min_amount_token_out {
+                // throw early if we cannot support this swap anyway due to liquidity being too low
+                return Err(DexError::NotEnoughLiquidityOf(token_out));
+            }
+
+            let amount_token_out = Self::out_given_in(
+                amount_token_in,
+                balance_token_in,
+                balance_token_out,
+                self.swap_fee_percentage,
+            )?;
+
+            if balance_token_out < amount_token_out {
+                // liquidity too low
+                return Err(DexError::NotEnoughLiquidityOf(token_out));
+            }
+
+            if amount_token_out < min_amount_token_out {
+                // thrown if too much slippage occured before this tx gets executed
+                // as a sandwitch atack prevention
+                return Err(DexError::TooMuchSlippage);
+            }
+
+            // transfer token_in from user to the contract
+            self.transfer_from_tx(token_in, caller, this, amount_token_in)??;
+            // transfer token_out from contract to user
+            self.transfer_tx(token_out, caller, amount_token_out)??;
+
+            // emit event
+            Self::emit_event(
+                self.env(),
+                Event::Swapped(Swapped {
+                    caller,
+                    token_in,
+                    token_out,
+                    amount_in: amount_token_in,
+                    amount_out: amount_token_out,
+                }),
+            );
+
+            Ok(())
+        }
+
+        /// Liquidity deposit
+        ///
+        /// Can only be performed by an account with a LiquidityProvider role
+        /// Caller needs to give at least the passed amount of allowance to the contract to spend the deposited tokens on his behalf
+        // Will revert if not enough allowance was given to the contract by the caller prior to executing this tx
+        #[ink(message)]
+        pub fn deposit(&mut self, deposits: Vec<(AccountId, Balance)>) -> Result<(), DexError> {
+            let this = self.env().account_id();
+            let caller = self.env().caller();
+
+            // check role, only designated account can add liquidity
+            <Self as AccessControlled>::check_role(
+                self.access_control,
+                caller,
+                Role::LiquidityProvider(this),
+                Self::cross_contract_call_error_handler,
+                Self::access_control_error_handler,
+            )?;
```

Comment:
All of the calls to `check_role` look almost exactly the same:
```rust
<Self as AccessControlled>::check_role(
                self.access_control,
                caller,
                <expected_role>,
                Self::cross_contract_call_error_handler,
                Self::access_control_error_handler,
            )?;
```
We could extract it to a method on `Self`:
```rust
fn check_role(expected: Role) {
    let caller = self.env().caller();
   <Self as AccessControlled>::check_role(
                self.access_control,
                caller,
                <expected_role>,
                Self::cross_contract_call_error_handler,
                Self::access_control_error_handler,
            )?;
}
```